### DISCUSSION TITLE
How do I make the custom created extension support Github Deployments with markdown_extensions?

### DISCUSSION THREAD

**Discussion by karl-police** [mkdocs/mkdocs#3492] [Q&A]
Continues https://github.com/mkdocs/mkdocs/discussions/3491

I have this in my mkdocs.yml

```yml
markdown_extensions:
  - docs.extensions.func_args
  - docs.extensions.func_returns
 ```

It runs well on localhost, but it doesn't on Github Deployments, it says "docs" not found

running with ``mkdocs gh-deploy --force``

while I am using ``mkdocs serve`` to locally test it

help

here's a deploy log

https://github.com/karl-police/mkdocs-testing/actions/runs/7026624709/job/19119681637

**Comment by kamilkrzyskow** [comment]
The `docs` module is not found by the importlib, most likely it's not in any of the `sys.path` directories.
There are multiple possible reasons for this:
1. Lack of or badly configured PYTHONPATH or PYTHONHOME
https://stackoverflow.com/questions/56047036/from-x-import-y-works-on-windows-but-not-on-ubuntu
https://stackoverflow.com/questions/7850908/what-exactly-should-be-set-in-pythonpath
2. Lack of the `__init__.py` file inside of the `docs` directory 
https://unix.stackexchange.com/questions/97709/no-module-found-error-in-linux-but-works-fine-in-windows

~~You could test it manually with `python -c "import sys; print(sys.path)"` and `python -c "import docs.extensions.func_args"` in the CI~~
Read below

**Reply by karl-police** [reply]
I tried adding the pwd path to both and more __init__.py but it didn't work

**Reply by karl-police** [reply]
Well, on my end I actually never ran ``mkdocs serve`` I ran ``python -m mkdocs serve``.

When I changed the command to ``- run: python -m mkdocs gh-deploy --force`` it worked and deployed don't know why

**Reply by kamilkrzyskow** [reply]
I did some manual testing, and yeah my bad on the `You could test it manually` part. Here is a workflow run of my findings, I failed it explicitly in the `on_post_build` event, so pay no mind to the red error.

https://github.com/kamilkrzyskow/mkdocs-ci-issue/actions/runs/7033676577/job/19140044183#step:7:1
I forgot to mention before, this is how you add an environmental variable in the GitHub Workflow.
`echo "PYTHONPATH=/home/runner/work/mkdocs-testing/mkdocs-testing/" >> $GITHUB_ENV`

https://github.com/kamilkrzyskow/mkdocs-ci-issue/actions/runs/7033676577/job/19140044183#step:8:13
`python -c "import sys; print(sys.path)"` resulted in a list of paths including an empty `''` string, which I'm assuming automatically points to the current working directory `pwd`. 

https://github.com/kamilkrzyskow/mkdocs-ci-issue/actions/runs/7033676577/job/19140044183#step:9:1
`python -c "import docs.extensions.func_args"` runs successfully after adding the PYTHONPATH, but due to your result I'm guessing that it's not needed when using `python`, as the empty `''` string points to `pwd`.

As for the question why doesn't it work when you run `mkdocs` instead of `python ...`,
https://github.com/kamilkrzyskow/mkdocs-ci-issue/actions/runs/7033676577/job/19140044183#step:10:139
I printed here the `sys.path` in separate lines and when running `mkdocs` the empty `''` string is replaced with `/opt/hostedtoolcache/Python/3.12.0/x64/bin`, so most certainly it didn't work on your end because it lacked the proper PYTHONPATH.

I will remove the fork in a few days or so, but I will give you some time to analyse it if you want. I debugged the `sys.path` from inside the `mkdocs` using a hook, a handy trick for future issues ✌️ 

> Note that I did add the `__init__.py` file, but as you didn't add it then, I guess, it can be omitted in this case.

**Reply by karl-police** [reply]
Idk I used ``python -m mkdocs gh-deploy --force`` on Github and that just solved everything

**Reply by kamilkrzyskow** [reply]
Yes, the change of command allowed you to deploy successfully 👍. However, results aren't everything, and you didn't solve the exact issue of deploying with the `mkdocs gh-deploy --force` command and I explained why that was in the comment.

**Reply by karl-police** [reply]
Is ``echo "PYTHONPATH=/home/runner/work/mkdocs-testing/mkdocs-testing/" >> $GITHUB_ENV`` the right way to add the path or are there other ways? Because it's using echo. What happens to the already set PYTHONPATH? And what even is that? Is it a path to the Python Binary?

**Reply by kamilkrzyskow** [reply]
https://docs.python.org/3/using/cmdline.html#environment-variables
`PYTHONPATH` is an environmental variable. As far, as I've seen `PYTHONPATH` isn't set on the GitHub CI worker. You can check set environmental variables yourself via expanding the run command. 

`$GITHUB_ENV` refers to a file, which contains environmental variables that GitHub loads before each command
`>>` appends the result of the previous command to the end of that file

To extend an already created PYTHONPATH, I guess you could do, but I haven't tried it:
`"PYTHONPATH=$PYTHONPATH:/home/runner/work/mkdocs-testing/mkdocs-testing/"`

<details>
<summary>Screenshot</summary>

![image](https://github.com/mkdocs/mkdocs/assets/34622465/580ce417-1340-44b4-b63c-842e641d717b)

</details>

**Reply by redtide** [reply]
> Is `echo "PYTHONPATH=/home/runner/work/mkdocs-testing/mkdocs-testing/" >> $GITHUB_ENV` the right way to add the path or are there other ways?

When I do that question to myself when working with new things, the answer is usually to search those terms in a search engine and check if the results returns the official [docs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable), otherwise in [Stack Overflow](https://stackoverflow.com/a/70419338) :)
